demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c

```diff
@@ -37,72 +46,240 @@ static AzureIoTResult_t prvBase64Decode( uint8_t * base64Encoded,
+     *
+     * If an image doesn't evenly divide into chunks of size azureiotflashL475_FLASH_ROW_SIZE,
+     * we write the leftovers by the double word. If the image evenly divides, we skip this.
+     *
+     */
+    uint8_t * pucNextWriteAddr = pxAduImage->xUpdatePartition + ulOffset;
+    uint8_t * pucNextReadAddr = pData;
+    AzureIoTResult_t xResult = eAzureIoTSuccess;
+    bool xIsLastBlock = pxAduImage->ulImageFileSize - ulOffset <= ulBlockSize;
+
+    /* no need to use double word programing at the end if the image fits evenly into chunks. */
+    bool xImageDividesEvenly = pxAduImage->ulImageFileSize % azureiotflashL475_FLASH_ROW_SIZE == 0;
+
+    /* starting address of last full sized chunk (of the whole image) - may not be in the last block */
+    uint8_t * pucLastChunkWriteAddr = pxAduImage->xUpdatePartition + pxAduImage->ulImageFileSize - ( ( xImageDividesEvenly ? 1 : 2 ) * azureiotflashL475_FLASH_ROW_SIZE );
+
+    /* ending address of the last full chunk in this block */
+    uint8_t * pucBlockEndAddr = pxAduImage->xUpdatePartition + ulOffset + ulBlockSize - ( ( xIsLastBlock && !xImageDividesEvenly ) ? azureiotflashL475_FLASH_ROW_SIZE : 0 );
+
+    HAL_FLASH_Unlock();
+
+    while( pucNextWriteAddr < pucBlockEndAddr )
+    {
+        /* For the last full chunk of the image written to the device (at address pucLastChunkWriteAddr), use FLASH_TYPEPROGRAM_FAST_AND_LAST */
+        if( HAL_FLASH_Program( ( pucNextWriteAddr >= pucLastChunkWriteAddr ) ?
+                               FLASH_TYPEPROGRAM_FAST_AND_LAST : FLASH_TYPEPROGRAM_FAST, ( uint32_t ) pucNextWriteAddr, ( uint32_t ) pucNextReadAddr ) != HAL_OK )
+        {
+            /* Error occurred while writing data in Flash memory */
+            xResult = eAzureIoTErrorFailed;
+            break;
+        }
+
+        pucNextWriteAddr += azureiotflashL475_FLASH_ROW_SIZE;
+        pucNextReadAddr += azureiotflashL475_FLASH_ROW_SIZE;
+    }
 
-    // TODO: Fill in to write pData of size ulBlockSize to ulOffset in memory.
+    /* Write any leftover data that didn't evenly fit into the chunks */
+    if( xIsLastBlock && ( pucNextWriteAddr > pucLastChunkWriteAddr + azureiotflashL475_FLASH_ROW_SIZE ) )
+    {
+        while( pucNextWriteAddr < pucBlockEndAddr + azureiotflashL475_FLASH_ROW_SIZE )
+        {
+            /* Program double words until the end of the file */
+            if( HAL_FLASH_Program( FLASH_TYPEPROGRAM_DOUBLEWORD, ( uint32_t ) pucNextWriteAddr, ( uint64_t ) *( uint32_t * ) pucNextReadAddr | ( ( uint64_t ) *( uint32_t * ) ( pucNextReadAddr + 4 ) ) << 32 ) != HAL_OK )
+            {
+                /* Error occurred while writing data in Flash memory */
+                xResult = eAzureIoTErrorFailed;
+                break;
+            }
+
+            pucNextWriteAddr += azureiotflashL475_DOUBLE_WORD_SIZE;
+            pucNextReadAddr += azureiotflashL475_DOUBLE_WORD_SIZE;
+        }
```

Comment:
So, this was quite an adventure to learn about, because doing this would completely crash the device for seemingly no reason! We can't do this because of [data alignment requirements](https://en.wikipedia.org/wiki/Data_structure_alignment). To test this, the device crashed when I read `*( uint64_t * )(pucNextReadAddr)`, but doesn't crash when I access `*( uint64_t * )(pucNextReadAddr+4)`

Also interesting is that this:
```
uint64_t expandedInVariable= ( uint64_t ) *( uint32_t * ) pucNextReadAddr | ( ( uint64_t ) *( uint32_t * ) ( pucNextReadAddr + 4 ) ) << 32;
AZLogInfo( ( "32_t words expanded: %016x\n", ( uint64_t ) *( uint32_t * ) pucNextReadAddr | ( ( uint64_t ) *( uint32_t * ) ( pucNextReadAddr + 4 ) ) << 32 ) );
AZLogInfo( ( "expandedInVariable: %016x\n", expandedInVariable));
AZLogInfo( ( "expandedInVariable cast to uint64_t: %016x\n", (uint64_t)expandedInVariable  ));
AZLogInfo( ( "first word in programming function: %016x\n", (uint32_t)expandedInVariable) );
AZLogInfo( ( "second word in programming function: %016x\n", (uint32_t)(expandedInVariable>> 32)  ) );
```

prints out this:
```
32_t words expanded: 00000000f102fa01
expandedInVariable: 0000000000000000
expandedInVariable cast to uint64_t: 0000000000000000
first word in programming function: 00000000f102fa01
second word in programming function: 00000000fc0cfa20
```
but that I still have no clue why we get the 2nd and 3rd lines as all 0000, since it doesn't behave that way for me in an online C sandbox 🤷‍♀️

Reply:
The suggestion worked beautifully in the online sandbox though :(

Reply:
Better explanation of why data alignment requirements crashed the device:

> On some systems, an int value must be stored in an address that's a multiple of 4 (or 2, or 8). This is called an [alignment](http://en.wikipedia.org/wiki/Data_structure_alignment) requirement. Depending on whether the address of c happens to be properly aligned or not, the program may crash.

https://stackoverflow.com/questions/17260527/what-are-the-rules-for-casting-pointers-in-c

Reply:
I have a feeling the printing has to do with limited functionality for embedded devices. 

https://eevblog.com/forum/microcontrollers/printf-uint64_t-with-arm-gcc/

I wonder if the buffer declaration as a `uint8_t` makes it unaligned? Not sure since it's been a while since I had to deal with misaligned memory with embedded devices. Just out of curiosity, I wonder if the buffer is declared as a uint64_t if it aligns properly?

Reply:
we can set [pucOutDataPtr](https://github.com/Azure-Samples/iot-middleware-freertos-samples/blob/a242c0604b6e41ae9c60fec9c089547917e36367/demos/sample_azure_iot_adu/sample_azure_iot_adu.c#L426) to be a uint64_t * that can be dereferenced (although we have to initialize the value of it to something to do this and then assign it to a char pointer, which all feels a little hacky). However, we send a pointer to this pointer to AzureIoTHTTP_Request, which then puts a new address in pucOutDataPtr that's defined as a char pointer, so after this function call, pucOutDataPtr is an address that isn't necessarily aligned. We could modify the functionality in AzureIoTHTTP_Request to use uint64_t data, but I'm guessing there would probably be other ramifications to that that would be larger than just bit shifting the last pieces of data in the new bin file.

So I'm going to say the best course of action is to leave this as is!